desktop-widgets/importgps.cpp

```diff
@@ -0,0 +1,294 @@
+		LocationUI->diveSiteCoordinates->editingFinished();
+	} else {
+		close();
+	}
+}
+
+//  Read text from the present position in the file until
+//  the first 'delim' character is encountered.
+int ImportGPS::getSubstring(QFile *file, QString *bufptr, char delim)
+{
+	char c;
+	bufptr->clear();
+	do {
+		if (file->read(&c, 1) <= 0) // EOF
+			return 1;
+		if ( c == delim) break;
+		bufptr->append(QChar(c));
+	} while (c != delim);
+	return 0;
+}
+
+// Find the next occurence of a specified target GPX element in the file,
+// characerised by a "<xxx " or "<xxx>" character sequence.
+// 'target' specifies the name of the element searched for.
+// termc is the ending character of the element name search = '>' or ' '.
+int ImportGPS::findXmlElement(QFile *fileptr, QString target, QString *bufptr, char termc)
+{
+	bool match = false;
+	char c;
+	char skipdelim = (termc == ' ') ? '>' : ' ';
+	do {    // Skip input until first start new of element (<) is encountered:
+		if (getSubstring(fileptr, bufptr, '<'))
+			return 1;  // EOF
+		bufptr->clear();
+		do {  // Read name of following element and store it in buf
+			if (fileptr->read(&c, 1) <= 0)  // EOF encountered
+				return 1;
+			if ((c == '>') || (c == ' '))   // found a valid delimiter
+				break;
+			bufptr->append(QChar(c));
+		} while ((c != '>') && (c != ' '));
+		if (*bufptr == "/trk") // End of GPS track found: return EOF
+			return 1;
+		if (c == skipdelim)
+			continue;  // if wrong delimiter for this element was found, redo from start
+		if (*bufptr == target)       // Compare xml element name from gpx file with the
+			match = true;    // the target element searched for.
+	} while (match == false);
+	return 0;
+}
+
+// Find the coordinates at the time specified in coords.start_dive
+// by searching the gpx file "fileName". Here is a typical trkpt element in GPX:
+// <trkpt lat="-26.84" lon="32.88"><ele>-53.7</ele><time>2017-08-06T04:56:42Z</time></trkpt>
+int ImportGPS::getCoordsFromFile()
+{
+	struct tm tm1;
+	time_t when = 0;
+	double lon, lat;
+	int line = 0;
+	int64_t time_offset = coords.settingsDiff_offset + coords.timeZone_offset;
+	time_t divetime;
+	bool first_line = true;
+	bool found = false;
+	divetime = coords.start_dive;
+	QString buf;
+	QFile f1;
+	f1.setFileName(fileName);
+	if (!f1.open(QIODevice::ReadOnly | QIODevice::Text)) {
+		QByteArray local8bitBAString1 = fileName.toLocal8Bit();
+		char *fname = local8bitBAString1.data();   // convert QString to a C string filename
+		fprintf(stderr, "GPS file open error: file name = %s\n", fname);
+		return 1;
+	}
+
+#ifdef GPSDEBUG
+	struct tm time; // decode the time of start of dive:
+	utc_mkdate(divetime, &time);
+	int dyr,dmon,dday,dhr,dmin;
+	dyr = time.tm_year;
+	dmon = time.tm_mon;
+	dday = time.tm_mday;
+	dhr = time.tm_hour;
+	dmin = time.tm_min;
+#endif
+
+	do {
+		line++;  // this is the sequence number of the trkpt xml element processed
```

Comment:
this is again feels like doing some XML parsing that can be done with Qt objects.